core/src/main/java/juzu/impl/common/Tools.java

```diff
@@ -194,7 +197,11 @@ public boolean hasNext() {
 
           @Override
           public T next() {
-            if (!hasNext()) {
+            if(next != null) {
```

Comment:
Yep, it's necessary.
In fact, the second condition will use the "hasNext()" method to iterate over the original iterator and to assign the next element, that's why we have to check again.
For the first test "if(next != null)", this will check if the "next" item was consulted or not, if not, return its value and make it "null" to not return it again by calling "next()" method of specific iterator.